Network rewrite

Him188 opened the pull request.

Him188 committed:
New network: framework infrastructure

Him188 committed:
Add AtomicLazy

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/net/impl/netty/NettyNetworkHandler.kt:
```diff
@@ -0,0 +1,138 @@
+        val contextResult = CompletableDeferred<ChannelHandlerContext>()
+
+        val eventLoopGroup = NioEventLoopGroup()
+        val bootstrap = Bootstrap()
+        bootstrap.group(eventLoopGroup)
+            .channel(NioSocketChannel::class.java)
+            .handler(object : ChannelInitializer<SocketChannel>() {
+                override fun initChannel(ch: SocketChannel) {
+                    ch.pipeline().addLast(object : ChannelInboundHandlerAdapter() {
+                        override fun channelActive(ctx: ChannelHandlerContext) {
+                            contextResult.complete(ctx)
+                            setupChannelContext(ctx)
+                        }
+                    })
+                }
+            })
+        bootstrap.connect(address)
```
1. 根据 Java Netty 习惯应该直接 `Bootstrap().......connect()` 而不需要单独 `val bootstrap`
2. `EventLoopGroup` 不应该每次都 `new`, 一个 `Bot` 生命周期应该使用同一个 `EventLoopGroup`
3.
```java
                .connect()
                .addListener(f -> {
                    if (!f.isSuccess()) {
                        contextResult.completeExceptionally(f.cause());
                    }
                })
```

Karlatemp replied:
附加: `Bot.close()` 时应该 `EventLoopGroup.shutdown() / shutdownGracefully()`

Him188 replied:
我先留个 todo , bot那边也要重构才能写这个

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/net/impl/netty/NetyNetworkHandlerFactory.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+internal object PacketDecodeInboundHandler : ChannelInboundHandlerAdapter() {
+    override fun channelRead(ctx: ChannelHandlerContext?, msg: Any?) {
+        super.channelRead(ctx, msg)
+    }
+}
```
Decoder 最好继承 `ByteToMessageDecoder`

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/net/impl/netty/NetyNetworkHandlerFactory.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+internal object PacketDecodeInboundHandler : ChannelInboundHandlerAdapter() {
```
`object` 在 Netty 中会有 `Sharable` 的问题
```java
public abstract class ChannelHandlerAdapter implements ChannelHandler {

    // Not using volatile because it's used only for a sanity check.
    boolean added;

    /**
     * Throws {@link IllegalStateException} if {@link ChannelHandlerAdapter#isSharable()} returns {@code true}
     */
    protected void ensureNotSharable() {
        if (isSharable()) {
            throw new IllegalStateException("ChannelHandler " + getClass().getName() + " is not allowed to be shared");
        }
    }

    /**
     * Return {@code true} if the implementation is {@link Sharable} and so can be added
     * to different {@link ChannelPipeline}s.
     */
    public boolean isSharable() {
        /**
         * Cache the result of {@link Sharable} annotation detection to workaround a condition. We use a
         * {@link ThreadLocal} and {@link WeakHashMap} to eliminate the volatile write/reads. Using different
         * {@link WeakHashMap} instances per {@link Thread} is good enough for us and the number of
         * {@link Thread}s are quite limited anyway.
         *
         * See <a href="https://github.com/netty/netty/issues/2289">#2289</a>.
         */
        Class<?> clazz = getClass();
        Map<Class<?>, Boolean> cache = InternalThreadLocalMap.get().handlerSharableCache();
        Boolean sharable = cache.get(clazz);
        if (sharable == null) {
            sharable = clazz.isAnnotationPresent(Sharable.class);
            cache.put(clazz, sharable);
        }
        return sharable;
    }
```

Him188 commented:
TODOs:

- [x] bot close 时抛出 `BotOfflineEvent.Active`
- [x] 登录完成 `BotOfflineEvent`
- [x] 重连完成 `BotReloginEvent`
----
- [x] heartbeat
- [x] `registerClientOnline()`
- [x] skey refresh

Him188 committed:
Remove rubbish; fix build; add todos

Him188 committed:
Buildscript: Fix test artifacts included in main sourceset

Him188 committed:
Set state to `StateClosed` appropriately and update docs

Karlatemp committed:
NettyTestUnit

Him188 committed:
Relocates netty files to correct packages

Him188 committed:
Move EMPTY_BYTE_ARRAY to mirai-core-utils

Him188 committed:
Migrate network logger and old sendWithoutExpect and sendAndExpect

Him188 committed:
Implement NetworkHandlerSelector, serverList, and implement them into Bot

Him188 committed:
Add AbstractKeepAliveNetworkHandlerSelectorTest

Him188 committed:
Extract netty utils

Him188 committed:
Rearrange QQAndroidBot

Him188 committed:
Decouple SSO login processor

Him188 committed:
Make QQAndroidClient lighter

Him188 committed:
Add NetworkHandler.isOk to avoid `state == OK`

Him188 committed:
Add state observer

Him188 committed:
States and connection impl

Him188 committed:
Add mirai-core-compileKotlinJvm.txt and mirai-core-compileKotlinAndroid.txt

Him188 committed:
Rearrange packages

Him188 committed:
ComponentStorage

Him188 committed:
Build: add slf4j

Him188 committed:
ContactUpdater and ContactCacheService

Him188 committed:
Integrate ServerList into component system

Him188 committed:
fix build

Him188 committed:
Continuing implement states for NetworkHandler

Him188 committed:
Prefer generified outgoing packets in all factories

Him188 committed:
Rearrange packages for component system

Him188 committed:
IDEA removed my import

Him188 committed:
Processors

Him188 committed:
Fix LongMessageInternal missing parameter (due to merge)

Him188 committed:
Fix test KeepAliveNetworkHandlerSelectorTest

Him188 committed:
Do not resume state while it is already closed

Him188 committed:
BotOnlineEvent, BotReloginEvent, BotOfflineEvent

Him188 commented:
- [x] consider remove state CONNECTION_LOST
- [x] unit tests for events
- [x] selector: support max attempts limit 
- [x] unit tests for state transition for `impl.netty`:
  - [x] INITIALIZED -> CONNECTING
  - [x] CONNECTING -> OK
  - [x] OK -> CONNECTING
  - [x] CONNECTING -> CLOSED
  - [x] OK -> CLOSED
- [x] production test for system hibernation
- [x] production test for unavailable Internet connection
- [x] review `BotConfiguration` and deprecate unused items
- [x] thorough self-review

Him188 committed:
Remove state CONNECTION_LOST

Him188 committed:
Add utils for tests

Him188 commented:
- ~支持记录模式，用于调试 (可以推迟到之后版本)~

Him188 committed:
Rearrange network tests

Him188 committed:
Components misc implementations including (ABI changes):
- ComponentStorage and its impls
- ContactCacheService
- ContactUpdater
- HeartbeatProcessor
- SsoProcessor
- QQABot

Him188 commented:
- [x] 检查 net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler.StateConnecting 会不会因为无网络导致的无限循环收集异常而内存占用越来越大
- [x] 检查 #1214

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/component/ComponentKey.kt:
```diff
@@ -0,0 +1,68 @@
+
+        private fun KType.renderType(fullName: Boolean) = classifier.renderClassifier(fullName)
+
+        private fun KTypeParameter.renderTypeParameter(fullName: Boolean): String {
+            val upperBounds = upperBounds
+            return when (upperBounds.size) {
+                0 -> toString()
+                1 -> "ComponentKey<${upperBounds[0].renderType(fullName)}>"
+                else -> "ComponentKey<${upperBounds.joinToString(" & ") { it.renderType(fullName) }}>"
+            }
+        }
```
Excepted?

`smartToString` -> `componentName` -> `renderClassifier` -> `renderTypeParameter`

`"ComponentKey<ComponentKey<.......>>`

Him188 committed:
Add NetworkHandlerFactory and tests for NetworkHandler

Karlatemp commented on mirai-core/src/commonTest/kotlin/network/component/ComponentStorageTest.kt:
```diff
@@ -0,0 +1,84 @@
+    companion object : ComponentKey<TestComponent3>
+}
+
+internal class ComponentStorageTest : AbstractTest() {
```
`CombinedComponentStorage` test?

Him188 committed:
CombinedStorageTest

Him188 committed:
fix componentName and smartToString

Him188 committed:
BotOfflineEventMonitor

Him188 committed:
Extract sync logic

Him188 committed:
Implement lifecycle management for network

Him188 committed:
api dump

Him188 committed:
Test: state tests

Him188 committed:
Test: ExceptionCollectorTest

Him188 committed:
Test: NettyEndlessReconnectionTest

Him188 committed:
Test: network scope closed on bot close

Him188 committed:
Close bot if first login failed

Him188 committed:
Fix state resume

Him188 committed:
Fix sso and event logging

Him188 committed:
Improve logging

Him188 committed:
Remove all unstable kotlin.Duration usages

Him188 committed:
Properly suspend coroutine is not available for send

Him188 committed:
Support network selector retry

Him188 committed:
Review BotConfiguration and implement relevant configs, implement alive heartbeat, fix behavior on resume

Him188 commented:
- [x] ConfigPushSvc.PushReq

Him188 committed:
Fix and improve event logging

Him188 committed:
Add suppress for ExperimentalSerializationApi

Him188 committed:
Extract some logic out of QQAndroidBot

Him188 committed:
Remove notes that was for rewrite

Him188 commented:
## For reviewer
*(anyone is welcomed, if any)*
### 主要修改

#### main

- class `net.mamoe.mirai.internal.AbstractBot`
- class `net.mamoe.mirai.internal.QQAndroidBot`
- package `net.mamoe.mirai.internal.network.handler`
- package `net.mamoe.mirai.internal.network.context`
- package `net.mamoe.mirai.internal.network.component`
- package `net.mamoe.mirai.internal.network.components`
- package `net.mamoe.mirai.internal.network.impl.netty`

#### test
- package `net.mamoe.mirai.internal.network`

### 建议阅读顺序:

- class `net.mamoe.mirai.internal.AbstractBot`
- class `net.mamoe.mirai.internal.QQAndroidBot`

## 架构总览

### `NetworkHandler`

<!--
 https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiY2xhc3NEaWFncmFtXG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyU3VwcG9ydFxuXG5jbGFzcyBOZXR3b3JrSGFuZGxlclxuXG5jbGFzcyBOZXR0eU5ldHdvcmtIYW5kbGVyXG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyRmFjdG9yeVxuXG5jbGFzcyBOZXR0eU5ldHdvcmtIYW5kbGVyRmFjdG9yeVxuXG5OZXR3b3JrSGFuZGxlciA8LS0gTmV0d29ya0hhbmRsZXJGYWN0b3J5ICA6IHByb2R1Y2VzXG5cbk5ldHdvcmtIYW5kbGVyRmFjdG9yeSA8fC0tIE5ldHR5TmV0d29ya0hhbmRsZXJGYWN0b3J5IDogaW1wbGVtZW50c1xuXG5OZXR0eU5ldHdvcmtIYW5kbGVyIDwuLiBOZXR0eU5ldHdvcmtIYW5kbGVyRmFjdG9yeSA6IHByb2NldWRlc1xuXG5jbGFzcyBTZWxlY3Rvck5ldHdvcmtIYW5kbGVyIHtcbiAtIHNlbGVjdG9yOiBOZXR3b3JrSGFuZGxlclNlbGVjdG9yfip-XG59XG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyU2VsZWN0b3J-SH4ge1xuICBIICtnZXRSZXN1bWVkSW5zdGFuY2UoKSpcbiAgSCArYXdhaXRSZXN1bWVJbnN0YW5jZSgpKlxufVxuXG5OZXR3b3JrSGFuZGxlciA8fC0tIE5ldHdvcmtIYW5kbGVyU3VwcG9ydFxuTmV0d29ya0hhbmRsZXJTdXBwb3J0IDx8LS0gTmV0dHlOZXR3b3JrSGFuZGxlclxuTmV0d29ya0hhbmRsZXIgPHwtLSBTZWxlY3Rvck5ldHdvcmtIYW5kbGVyXG5cblNlbGVjdG9yTmV0d29ya0hhbmRsZXIgby0tIE5ldHdvcmtIYW5kbGVyU2VsZWN0b3IgOiBzdHJhdGVneVxuXG5OZXR3b3JrSGFuZGxlclNlbGVjdG9yIDx8LS0gQWJzdHJhY3RLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yXG5cbmNsYXNzIEFic3RyYWN0S2VlcEFsaXZlTmV0d29ya0hhbmRsZXJTZWxlY3RvciB7XG4gICNjcmVhdGVJbnN0YW5jZSgpXG59XG5cbmNsYXNzIEZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIHtcbiAgLWZhY3Rvcnk6IE5ldHdvcmtIYW5kbGVyRmFjdG9yeVxufVxuXG5BYnN0cmFjdEtlZXBBbGl2ZU5ldHdvcmtIYW5kbGVyU2VsZWN0b3IgPHwtLUZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIDogZXh0ZW5kc1xuXG5OZXR3b3JrSGFuZGxlckZhY3RvcnkgLi5vIEZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIDogY29udGFpbnMiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ
-->

[![](https://mermaid.ink/img/eyJjb2RlIjoiY2xhc3NEaWFncmFtXG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyU3VwcG9ydFxuXG5jbGFzcyBOZXR3b3JrSGFuZGxlclxuXG5jbGFzcyBOZXR0eU5ldHdvcmtIYW5kbGVyXG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyRmFjdG9yeVxuXG5jbGFzcyBOZXR0eU5ldHdvcmtIYW5kbGVyRmFjdG9yeVxuXG5OZXR3b3JrSGFuZGxlciA8LS0gTmV0d29ya0hhbmRsZXJGYWN0b3J5ICA6IHByb2R1Y2VzXG5cbk5ldHdvcmtIYW5kbGVyRmFjdG9yeSA8fC0tIE5ldHR5TmV0d29ya0hhbmRsZXJGYWN0b3J5IDogaW1wbGVtZW50c1xuXG5OZXR0eU5ldHdvcmtIYW5kbGVyIDwuLiBOZXR0eU5ldHdvcmtIYW5kbGVyRmFjdG9yeSA6IHByb2NldWRlc1xuXG5jbGFzcyBTZWxlY3Rvck5ldHdvcmtIYW5kbGVyIHtcbiAtIHNlbGVjdG9yOiBOZXR3b3JrSGFuZGxlclNlbGVjdG9yfip-XG59XG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyU2VsZWN0b3J-SH4ge1xuICBIICtnZXRSZXN1bWVkSW5zdGFuY2UoKSpcbiAgSCArYXdhaXRSZXN1bWVJbnN0YW5jZSgpKlxufVxuXG5OZXR3b3JrSGFuZGxlciA8fC0tIE5ldHdvcmtIYW5kbGVyU3VwcG9ydFxuTmV0d29ya0hhbmRsZXJTdXBwb3J0IDx8LS0gTmV0dHlOZXR3b3JrSGFuZGxlclxuTmV0d29ya0hhbmRsZXIgPHwtLSBTZWxlY3Rvck5ldHdvcmtIYW5kbGVyXG5cblNlbGVjdG9yTmV0d29ya0hhbmRsZXIgby0tIE5ldHdvcmtIYW5kbGVyU2VsZWN0b3IgOiBzdHJhdGVneVxuXG5OZXR3b3JrSGFuZGxlclNlbGVjdG9yIDx8LS0gQWJzdHJhY3RLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yXG5cbmNsYXNzIEFic3RyYWN0S2VlcEFsaXZlTmV0d29ya0hhbmRsZXJTZWxlY3RvciB7XG4gICNjcmVhdGVJbnN0YW5jZSgpXG59XG5cbmNsYXNzIEZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIHtcbiAgLWZhY3Rvcnk6IE5ldHdvcmtIYW5kbGVyRmFjdG9yeVxufVxuXG5BYnN0cmFjdEtlZXBBbGl2ZU5ldHdvcmtIYW5kbGVyU2VsZWN0b3IgPHwtLUZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIDogZXh0ZW5kc1xuXG5OZXR3b3JrSGFuZGxlckZhY3RvcnkgLi5vIEZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIDogY29udGFpbnMiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiY2xhc3NEaWFncmFtXG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyU3VwcG9ydFxuXG5jbGFzcyBOZXR3b3JrSGFuZGxlclxuXG5jbGFzcyBOZXR0eU5ldHdvcmtIYW5kbGVyXG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyRmFjdG9yeVxuXG5jbGFzcyBOZXR0eU5ldHdvcmtIYW5kbGVyRmFjdG9yeVxuXG5OZXR3b3JrSGFuZGxlciA8LS0gTmV0d29ya0hhbmRsZXJGYWN0b3J5ICA6IHByb2R1Y2VzXG5cbk5ldHdvcmtIYW5kbGVyRmFjdG9yeSA8fC0tIE5ldHR5TmV0d29ya0hhbmRsZXJGYWN0b3J5IDogaW1wbGVtZW50c1xuXG5OZXR0eU5ldHdvcmtIYW5kbGVyIDwuLiBOZXR0eU5ldHdvcmtIYW5kbGVyRmFjdG9yeSA6IHByb2NldWRlc1xuXG5jbGFzcyBTZWxlY3Rvck5ldHdvcmtIYW5kbGVyIHtcbiAtIHNlbGVjdG9yOiBOZXR3b3JrSGFuZGxlclNlbGVjdG9yfip-XG59XG5cbmNsYXNzIE5ldHdvcmtIYW5kbGVyU2VsZWN0b3J-SH4ge1xuICBIICtnZXRSZXN1bWVkSW5zdGFuY2UoKSpcbiAgSCArYXdhaXRSZXN1bWVJbnN0YW5jZSgpKlxufVxuXG5OZXR3b3JrSGFuZGxlciA8fC0tIE5ldHdvcmtIYW5kbGVyU3VwcG9ydFxuTmV0d29ya0hhbmRsZXJTdXBwb3J0IDx8LS0gTmV0dHlOZXR3b3JrSGFuZGxlclxuTmV0d29ya0hhbmRsZXIgPHwtLSBTZWxlY3Rvck5ldHdvcmtIYW5kbGVyXG5cblNlbGVjdG9yTmV0d29ya0hhbmRsZXIgby0tIE5ldHdvcmtIYW5kbGVyU2VsZWN0b3IgOiBzdHJhdGVneVxuXG5OZXR3b3JrSGFuZGxlclNlbGVjdG9yIDx8LS0gQWJzdHJhY3RLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yXG5cbmNsYXNzIEFic3RyYWN0S2VlcEFsaXZlTmV0d29ya0hhbmRsZXJTZWxlY3RvciB7XG4gICNjcmVhdGVJbnN0YW5jZSgpXG59XG5cbmNsYXNzIEZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIHtcbiAgLWZhY3Rvcnk6IE5ldHdvcmtIYW5kbGVyRmFjdG9yeVxufVxuXG5BYnN0cmFjdEtlZXBBbGl2ZU5ldHdvcmtIYW5kbGVyU2VsZWN0b3IgPHwtLUZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIDogZXh0ZW5kc1xuXG5OZXR3b3JrSGFuZGxlckZhY3RvcnkgLi5vIEZhY3RvcnlLZWVwQWxpdmVOZXR3b3JrSGFuZGxlclNlbGVjdG9yIDogY29udGFpbnMiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)

`NetworkHandler` (简称 `NH`) 主要层次结构为 `NH` <-- `NHSupport` <-- `NettyNH`. 其中 `NHSupport` 是 `NH` 的基础实现. 而 `NettyNH` 则是使用 Netty 的特定实现.

一个 `NH` 维护一段连接. 在*简单断线*时会尝试重连 (下文详述).

`NHFactory` 承担稳定且通用的构造工作, 派生为 `NettyNHFactory` (生产 `NettyNH`).


`SelectorNH` 继承 `NH`, 通过 `NetworkHandlerSelector` (简称 `Selector`) 不断地构造 `NH` 实例, 完成状态切换, 并释放不可用的 `NH`. `SelectorNH` 为作为一个 proxy (delegate) 将对 `NH` 的请求转发给由 `Selector` 选择得到的实例.
`NH` 仅维护一段连接, 在*简单断线*时会尝试重连 (下文详述), 而*复杂或异常断线*则会关闭. `SelectorNH` 的职责就是有条理地处理 `NH` 的断线并构造新的实例替换.

每一个对象都只*专一地*做一件很小的事情. 是且仅是 `NettyNH` 会处理通过 `Netty` 建立的连接的断开 (可能是网络不稳定). 是且仅是 `Selector` 会维护 `NH` 内部出错和 `Bot` 被挤下线等异常的重连 (构造新的 `NH` 来避免产生复杂的状态关系).

### 实现

网络层由多个子模块 (组件) 组成, 分别负责一个特定的工作. 组件是 `Bot` 全局的. 各个组件除了对 `Bot`, `ComponentStorage` 外, 不持有其他状态量.

`NetworkHandler` 不持有对特定协议的引用. 而是调用通用的 `SsoProcessor`, `BotInitProcessor` 等组件.

`NetworkHandlerSupport` 将连接状态封装为 `BaseStateImpl`(简称 `State`). 各个状态互相独立, 拥有单独的 `CoroutineScope`. 状态的切换由旧状态自己进行而外部不能干涉. 切换时允许传递一些能在多个状态共享的数据.

在状态的基础上设计了 `StateObserver`, 用来监控状态的变化并允许为状态添加一些子任务. 有的 observer 会在合适的时机广播 `BotOnlineEvent` 等事件, 有的则通过日志记录变化以供调试.

----

设计上有很多细节均有注释标注. 而部分模块是沿用旧网络层的代码, 它可能仍然有较强耦合性, 但现阶段还没必要处理他们.

Him188 committed:
api dump

Him188 committed:
Set state to CLOSED on failure during connecting

Karlatemp commented on binary-compatibility-validator/android/api/binary-compatibility-validator-android.api:
```diff
@@ -2127,30 +2130,15 @@ public final class net/mamoe/mirai/event/events/BotOfflineEvent$MsfOffline : net
 	public fun getBot ()Lnet/mamoe/mirai/Bot;
+	public fun getCause ()Ljava/lang/Throwable;
 	public fun getReconnect ()Z
 	public fun hashCode ()I
```
Is it ok?

Him188 replied:
yes，@MiraiInternalAPI

Karlatemp commented on mirai-core-api/src/commonMain/kotlin/event/Event.kt:
```diff
@@ -139,17 +144,35 @@ public interface CancellableEvent : Event {
+    if (event is BotEvent) {
+        event.bot.logger.verbose { "Event: $event" }
+    } else {
+        MiraiLogger.TopLevel.verbose { "Event: $event" }
```
Maybe use `logger("EventPipeline")` will be better than `TopLevel`?

Karlatemp commented on mirai-core-utils/src/commonMain/kotlin/ExceptionCollector.kt:
```diff
@@ -0,0 +1,100 @@
+    @Volatile
+    private var last: Throwable? = null
+
+    @Synchronized
+    public fun collect(e: Throwable?) {
+        if (e == null) return
+        val last = last
+        if (last != null) {
+            last.itr().forEach { suppressed ->
+                if (suppressed.stackTrace.contentEquals(e.stackTrace)) {
+                    // filter out useless duplicates.
+                    return
+                }
+            }
+            e.addSuppressed(last)
+        }
+        this.last = e
+    }
```
只 判断 stackTrace 可能不太合理, `stackTrace` 一样而 `message` 可能会不样

为了执行效率考虑，也应该添加 `===` 短路判断

Him188 replied:
message 一般不重要，只要抛出的点一样就行

异常一般都是new了抛的（这里不是设计来去除 same instance），=== 不会 true

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/components/ConfigPushProcessor.kt:
```diff
@@ -0,0 +1,48 @@
+) : ConfigPushProcessor {
+    override suspend fun syncConfigPush(network: NetworkHandler) {
+        if (nextEventOrNull<ConfigPushSvc.PushReq.PushReqResponse>(60_000) == null) {
+            val bdhSyncer = network.context[BdhSessionSyncer]
```
`nextEventOrNull`: filtered bot?

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/components/SsoProcessor.kt:
```diff
@@ -0,0 +1,285 @@
+                            } else {
+                                WtLogin2.SubmitSliderCaptcha(client, ticket).sendAndExpect()
+                            }
+                        } else {
+                            // retry once
+                            if (!allowSlider) throw createUnsupportedSliderCaptchaException(allowSlider)
+                            allowSlider = false
+                            response = WtLogin9(client, allowSlider).sendAndExpect()
+                        }
```
此处最好的做法是断开当前连接重新完整连接一次, 而不是直接就把这个包发过去

Him188 replied:
should `allowSlider = false`

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/impl/netty/NettyNetworkHandler.kt:
```diff
@@ -0,0 +1,400 @@
+
+        override fun getCause(): Throwable? = collectiveExceptions.getLast()
+
+        override suspend fun sendPacketImpl(packet: OutgoingPacket): Boolean {
+            connection.await() // split line number
+                .writeAndFlush(packet)
+            return true
+        }
```
```kotlin
writeAndFlush().addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE)


writeAndFlush().addListener(ChannelFutureListener.CLOSE_ON_FAILURE)
```

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/impl/netty/NettyNetworkHandler.kt:
```diff
@@ -0,0 +1,400 @@
+        }
+
+        override suspend fun sendPacketImpl(packet: OutgoingPacket): Boolean {
+            connection.writeAndFlush(packet)
```
与上方 review 一样

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/impl/netty/nettyUtils.kt:
```diff
@@ -0,0 +1,67 @@
+}
+
+
+internal suspend inline fun joinCompleted(job: Job) {
+    if (job.isCompleted) job.join()
+}
```
Why it in `nettyUtils.kt`

Him188 replied:
暂时只used only in netty，放 mirai utils 可能没有必要

Karlatemp commented on mirai-core/src/commonMain/kotlin/utils/SubLogger.kt:
```diff
@@ -0,0 +1,63 @@
+    return SubLogger(name, this)
+}
+
+private class SubLogger(
+    private val name: String,
+    private val main: MiraiLogger,
+) : MiraiLoggerPlatformBase() {
+    override val identity: String? get() = main.identity
```
`override val isEnabled: Boolean get() = main.isEnabled`

Him188 committed:
Dispatch events in Bot scope

Him188 committed:
Add bot argument to PushReqResponse and filter in ConfigPushProcessorImpl

Him188 committed:
Use topLevelEventLogger instead of MiraiLogger.TopLevel for non-BotEvent events

Him188 committed:
Add writeAndFlushOrCloseAsync to handle exception on send properly

Him188 committed:
Improve logging in components

- Fix duplicated event logs
- Specify logger name for AccountSecretsManager
- Fix OtherClientUpdater logger
- Unify ServerList logger
- SubLogger.isEnabled

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/components/PacketLoggingStrategy.kt:
```diff
@@ -0,0 +1,100 @@
+            packet is Packet.NoLog -> {
+                // nothing to do
+            }
+            packet is Event && packet !is Packet.NoEventLog -> bot.logger.verbose {
+                "Event: $packet".replaceMagicCodes()
+            }
```
`Event.kt` 中 `broadcast` 会 `log`, 此处再 log 一次会重复 log

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/impl/netty/NettyNetworkHandler.kt:
```diff
@@ -0,0 +1,400 @@
+        }
+    }
+
+    protected open fun setupChannelPipeline(pipeline: ChannelPipeline, decodePipeline: PacketDecodePipeline) {
+        pipeline
+            .addLast(OutgoingPacketEncoder())
+            .addLast(LengthFieldBasedFrameDecoder(Int.MAX_VALUE, 0, 4, -4, 4))
+            .addLast(ByteBufToIncomingPacketDecoder())
+            .addLast(RawIncomingPacketCollector(decodePipeline))
+    }
+
+    // can be overridden for tests
+    protected open suspend fun createConnection(decodePipeline: PacketDecodePipeline): NettyChannel {
+        val contextResult = CompletableDeferred<NettyChannel>()
+        val eventLoopGroup = NioEventLoopGroup()
+
+        val future = Bootstrap().group(eventLoopGroup)
+            .channel(NioSocketChannel::class.java)
+            .option(ChannelOption.SO_KEEPALIVE, true)
+            .handler(object : ChannelInitializer<SocketChannel>() {
+                override fun initChannel(ch: SocketChannel) {
+                    ch.pipeline()
+                        .addLast(object : ChannelInboundHandlerAdapter() {
+                            override fun channelInactive(ctx: ChannelHandlerContext?) {
+                                eventLoopGroup.shutdownGracefully()
+                            }
+                        })
+
+                    setupChannelPipeline(ch.pipeline(), decodePipeline)
+                }
+            })
+            .connect(address)
+            .awaitKt()
```
看上去像是没添加异常处理器处理 Netty pipeline error, 以及网络 IO 错误

Him188 commented:
- [ ] 实现 `HeartbeatStrategy`

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/impl/netty/NettyNetworkHandler.kt:
```diff
@@ -0,0 +1,400 @@
+            return true
+        }
+
+        override suspend fun resumeConnection0() {
+            connectResult.join() // propagates exceptions
+            val connection = connection.await()
+            setState(this) { StateLoading(connection) }
+                ?.resumeConnection()
+                ?: this@NettyNetworkHandler.resumeConnection() // concurrently closed by other thread.
+        }
```
¿¿¿¿

https://scans.gradle.com/s/znkayqo4n6zpq/tests
```txt

Project | Tests | Outcome | Duration |  
-- | -- | -- | -- | --
  | :mirai-core | 151 passed, 1 failed | FAILED | 2m 5.451s |  
  | androidTest | 69 passed, 1 failed | FAILED | 1m 4.509s |  
  | net.mamoe.mirai.internal.network.impl.netty | 24 passed, 1 failed | FAILED | 57.591s |  
  | NettyHandlerEventTest | 10 passed, 1 failed | FAILED | 42.160s |  
  | BotOffline from CONNECTING TO CLOSED()



Exception in thread "DefaultDispatcher-worker-2 @coroutine#158" java.lang.NullPointerException
	at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler$StateConnecting.resumeConnection0(NettyNetworkHandler.kt:252)
	at net.mamoe.mirai.internal.network.handler.NetworkHandlerSupport$BaseStateImpl.resumeConnection(NetworkHandlerSupport.kt:157)
	at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler$StateConnecting$$special$$inlined$apply$lambda$1$1.invokeSuspend(NettyNetworkHandler.kt:226)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

Him188 commented on mirai-core/src/commonMain/kotlin/QQAndroidBot.kt:
```diff
@@ -66,6 +67,21 @@ internal open class QQAndroidBot constructor(
+     *
+     */
+    override val botReinitActions = if (debugConfiguration.allowReinitActions) {
+        BotReinitActionsImpl()
```
应该放 components

Him188 commented on mirai-core/src/commonMain/kotlin/network/components/BotReinitActions.kt:
```diff
@@ -0,0 +1,75 @@
+
+    fun addAction0(action: () -> Unit)
+
+    //@VisitOnlyForTests
```
`@TestOnly`

Him188 commented on mirai-core/src/commonTest/kotlin/network/framework/AbstractRealNetworkHandlerTest.kt:
```diff
@@ -65,27 +66,62 @@ internal abstract class AbstractRealNetworkHandlerTest<H : NetworkHandler> : Abs
+        login: (suspend NetworkHandler.() -> Unit)? = null
+    ) = withSsoProcessor(login = login, logout = null)
+
+    protected open fun withSsoProcessor(
```
这应该放到 sso test 里
Abstract 默认不带 sso 那个 component

Him188 commented on mirai-core/src/commonTest/kotlin/network/framework/AbstractRealNetworkHandlerTest.kt:
```diff
@@ -65,27 +66,62 @@ internal abstract class AbstractRealNetworkHandlerTest<H : NetworkHandler> : Abs
+    protected open val defaultComponents = ConcurrentComponentStorage().apply {
+        val components = this
+        val configuration = bot.configuration
+        set(BotReinitActions, BotReinitActions.byDelegate { bot.components[BotReinitActions] })
```
recursive call?
bot.components -> defaultComponents  -> bot.components

Karlatemp replied:
bot.components 不包含 defaultComponents

Him188 commented on mirai-core/src/commonMain/kotlin/QQAndroidBot.kt:
```diff
@@ -54,7 +54,8 @@ internal fun Bot.asQQAndroidBot(): QQAndroidBot {
 internal class BotDebugConfiguration(
-    var stateObserver: StateObserver? = LOGGING
+    var stateObserver: StateObserver? = LOGGING,
+    var allowReinitActions: Boolean = false,
```
不好, 应该用一个 component 由 tesst 覆盖而不是 bool 的开关

Him188 commented on mirai-core/src/commonMain/kotlin/QQAndroidBot.kt:
```diff
@@ -77,6 +93,11 @@ internal open class QQAndroidBot constructor(
                 private val shouldBroadcastRelogin = atomic(false)
+
+                init {
+                    botReinitActions.addAction { shouldBroadcastRelogin.value = false }
```
在这里操作类变量显得流程不清晰

Him188 commented on mirai-core/src/commonMain/kotlin/network/components/BotReinitActions.kt:
```diff
@@ -0,0 +1,75 @@
+import java.util.concurrent.ConcurrentLinkedQueue
+
+/**
+ * Reset bot and others components to initialize state for testing
```
看起来任何时候都不应该 reset component (增加的变化量容易引起意外的难以发现的情况并容易演变成旧网络层拥有的的复杂的状态), 而是要 new 一个新的 bot

sandtechnology commented:
目前新网络层无法正确处理离线事件：
- [ ] 在启动时接到MSFOffline->完全无响应
- [ ] 在启动完成时接到MSFOffline->显示掉线，后**0s**重连成功

Him188 committed:
Prototype EventDispatcher

Karlatemp committed:
Fix NPE in StateConnecting

Him188 committed:
Remove redundant properties

Him188 committed:
Fix docs

Karlatemp committed:
Exception Handling; Normal login tests

Him188 commented on mirai-core/src/commonTest/kotlin/network/impl/netty/NettyBotNormalLoginTest.kt:
```diff
@@ -0,0 +1,70 @@
+
+    @Test
+    fun `test login fail`() = runBlockingUnit {
+        resetSsoProcessor()
```
让 test instance 是 per method 是不是就不用 reset了？

Karlatemp replied:
bot instance? 则是, 目前测试架构是一个 test unit 公用一个 bot, 所以需要 reset

Him188 replied:
```
@TestInstance(PER_METHOD)
```
on class 呢

Karlatemp commented:
- [ ] 模拟网络不佳的情况
  - [ ] 中途断网
- [ ] 完整正常登入测试
  - [ ] 带 account secrets
    - [ ] account secrets 有效
    - [ ] account secrets 无效
  - [ ] 不带 account secrets
  - [ ] 非 *ANDROID_PHONE* 协议下的二次验证码解决机制

Him188 committed:
Add NHSupport.setState safe overload and fix tests

Him188 committed:
Fix imports

Him188 committed:
Add AbstractNettyNHTest.setSsoProcessor

Him188 committed:
Minor improvements after rebase

Him188 commented:
Rebase 到了 dev 并 squash 了一些相关 commits, 规范了 commit message.
